SignalServiceKit/src/Storage/OWSStorage.m

```diff
@@ -250,39 +160,69 @@ - (instancetype)initStorage
     self = [super init];
 
     if (self) {
-        if (![self tryToLoadDatabase]) {
-            // Failing to load the database is catastrophic.
-            //
-            // The best we can try to do is to discard the current database
-            // and behave like a clean install.
-            OWSProdCritical([OWSAnalyticsEvents storageErrorCouldNotLoadDatabase]);
-
-            // Try to reset app by deleting all databases.
-            //
-            // TODO: Possibly clean up all app files.
-            //            [OWSStorage deleteDatabaseFiles];
-
-            if (![self tryToLoadDatabase]) {
-                OWSProdCritical([OWSAnalyticsEvents storageErrorCouldNotLoadDatabaseSecondAttempt]);
-
-                // Sleep to give analytics events time to be delivered.
-                [NSThread sleepForTimeInterval:15.0f];
-
-                [NSException raise:OWSStorageExceptionName_NoDatabase format:@"Failed to initialize database."];
-            }
-        }
+        [self openDatabase];
+
+        [[NSNotificationCenter defaultCenter] addObserver:self
+                                                 selector:@selector(applicationDidEnterBackground:)
+                                                     name:OWSApplicationDidEnterBackgroundNotification
+                                                   object:nil];
+        [[NSNotificationCenter defaultCenter] addObserver:self
+                                                 selector:@selector(extensionHostWillEnterForeground:)
+                                                     name:OWSApplicationWillEnterForegroundNotification
+                                                   object:nil];
     }
 
     return self;
 }
 
+- (void)dealloc
+{
+    [[NSNotificationCenter defaultCenter] removeObserver:self];
+}
+
 - (StorageType)storageType
 {
     OWS_ABSTRACT_METHOD();
 
     return StorageType_Unknown;
 }
 
+- (void)openDatabase
+{
+    OWSAssertIsOnMainThread();
+
+    DDLogInfo(@"%@ %s", self.logTag, __PRETTY_FUNCTION__);
+
+    if (![self tryToLoadDatabase]) {
+        // Failing to load the database is catastrophic.
+        //
+        // The best we can try to do is to discard the current database
+        // and behave like a clean install.
+        OWSProdCritical([OWSAnalyticsEvents storageErrorCouldNotLoadDatabase]);
+
+        // Try to reset app by deleting all databases.
+        [OWSStorage deleteDatabaseFiles];
+
+        if (![self tryToLoadDatabase]) {
+            OWSProdCritical([OWSAnalyticsEvents storageErrorCouldNotLoadDatabaseSecondAttempt]);
+
+            // Sleep to give analytics events time to be delivered.
+            [NSThread sleepForTimeInterval:15.0f];
+
+            [NSException raise:OWSStorageExceptionName_NoDatabase format:@"Failed to initialize database."];
+        }
+    }
+}
+
+- (void)closeDatabase
+{
+    OWSAssertIsOnMainThread();
```

Comment:
No. First of all, there's a big difference between `OWSStorage` and `OWSSessionStorage`.  

 `OWSSessionStorage` is strict and closes its database & discards its connections - but only when:

* The app is in the background AND 
* Any open transactions have completed AND
* The app isn't processing incoming messages or sending messages in the background.  
   * There's some critical follow-up work TODOs around this third point. There's a card.

Note that `OWSStorage` doesn't do this.

As for explicitly invalidating connections, that would require the `OWSStorage` instances to retain a (presumably weak) reference to all open connections, and access to that collection would need to be synchronized.  I'm not sure it's worth the trouble.  If we ever tried to use an "invalid" connection, it should fail immediately since the database is closed.